### ISSUE TITLE
Doubled notifications

### ISSUE THREAD

**Issue by waclaw66** [vector-im/element-android#4152]
### Steps to reproduce

Just receive a message.

https://user-images.githubusercontent.com/15554561/135765548-a0284d91-fc22-40b4-a986-020245f7be76.mp4

### What did you expect?
Single notification for a message.

### What happened?
When single message is received two same notifications are shown for a little while in a row.
I tried to re-register pusher, create a new session, optimized init sync. Nothing helped.

### Your phone model

Xiaomi Mi 9T

### Operating system version

Android 11

### Application version and app store

1.3.1

### Homeserver

own

### Have you submitted a rageshake?

Yes

**Comment by ouchadam** [comment]
Hi @waclaw66 thanks for raising this, unfortunately I don't have a android 11 MIUI device to test on and I'm not familiar with their notification UI changes, would you be able to provide some more information?

- do you have the same issue when using a matrix.org account
- has this issue recently started happening, if so was it tied to an application update/homeserver upgrade
- do the messages stack as list with notification actions
- "notifications are shown for a little while in a row." does this mean the duplicate notification disappears 

| MIUI 12.0.3/Android 10 | Stock~ (Samsung One UI) Android 11 | 
| --- | --- |
![2021-10-04T13:01:44,528885888+01:00](https://user-images.githubusercontent.com/1848238/135847982-e7b98777-d8c4-410e-b388-7615ce93284b.png)|![2021-10-04T13:07:28,372840475+01:00](https://user-images.githubusercontent.com/1848238/135848668-3dbb6340-b276-418a-a939-ac194bf818b4.png)

**Comment by waclaw66** [comment]
@ouchadam:
some feedback...
- yes it does also using matrix.org account
- it happens a while, now I've tried version 1.2.2, it behaves the same
- messages stack same as on your screenshot, but for two messages there is count (4)
- yes, I see only single notification for a message, but it does a sound and counts twice

**Comment by ouchadam** [comment]
the video really helps clarify, thanks!

It's interesting to see the final expanded notification is 

```
test 1 
test 2
```

rather than 

```
test 1
test 1
test 2 
test 2
``` 

I'll keep investigating and see if I can reproduce on my side and hopefully fix

**Comment by ouchadam** [comment]
one last thing to confirm, does the issue happen for both notification styles? Settings -> Notifications -> Notifications Shade -> android / MIUI

![2021-10-04T20:29:03,743812769+01:00](https://user-images.githubusercontent.com/1848238/135912524-49e0e85f-532b-4679-9d86-b6fd7e333209.png)

when I select MIUI (Pocofone f1 MIUI 12.0.3.0) and turn off animations, I can see flashes of the summary group leaking through, there's also a few reports of the MIUI style having issues with other messaging apps (which I assume is a group summary issue)

**Comment by waclaw66** [comment]
It looks the same for both notification styles. Although I have observed that sometimes (quite seldom) the second notification counts up to 3 only.

**Comment by ouchadam** [comment]
Very interesting... the difference in network conditions also ties into my current race condition theory

My current thinking is that the way we update the android summary group notification (eg how gmail batches notifications under a single expandable notification) is updating out of sync to the individual message notifications, causing the group notification to display in the same place as the normal message, which in turn presents itself as 2 notifications for 1 message event (1 summary update, 1 message update).

We can see this happening in your video where the first message that is displayed is "$username: $message", this is a group summary and shouldn't be visible until there's multiple notifications to put in a group.

At the moment we create and display all the individual messages in a big loop, each iteration of the loop may take more or less time depending on fetching data, avatars, cancelling previous notifications etc then at the very end we update the group, however if we don't update the group fast enough sometimes wonky things can happen.

My plan is a big redesign of the notifications in order to create them all and then display them all in one go, hopefully eliminating the race conditions :crossed_fingers: 

There's also a lot of  redundant `cancelling -> showing` which might be causing issues which will also be addressed in the redesign.

tl;dr still working on it!

**Comment by waclaw66** [comment]
I think it is not completly resolved yet. I'm seeing it time to time even on my new Samsung device (S20 FE Android 11).  There is shown (2) for a single notification.

**Comment by ouchadam** [comment]
thanks for the update :+1:  when this occurs, does the notification expand to show multiple notifications, or is it still the 1 notification but notified twice?

also, would you say the issue occurs less often after upgrading to 1.3.7+?

**Comment by waclaw66** [comment]
There is shown only single notification but with badge (2). It does much less than before 1.3.7.
I'm not sure if it's related, but sometimes a notification doesn't dissapear when I click on it. Element is opened on that message, but notification keeps showing, it has to be swiped to side manually.

**Comment by ouchadam** [comment]
unfortunately this appears to be a samsung specific issue (possibly other OEMs), they're including any updates to the notification in their counter  

it's caused by the use of the conversations api and us updating existing messages, some times the messages are fetched in their final form but other times there could be edits, redactions or the server could give us more information about the message later on

for example, in screenshots above scenario I'm seeing the same message come through twice as we gain the sender information in separate updates 

```
sender - adam
message - a

sender - adam (@ouchadam:matrix.org)
message - a
```

the second notification is created to notify silently (no sound/vibration etc) but the samsung counter doesn't respect this. 

FWIW stock android doesn't include this counter

| ANDROID 11 | ANDROID 12 |
| --- | --- |
|![Screenshot_20220120_135907](https://user-images.githubusercontent.com/1848238/150353635-c116505f-97c8-40e0-b11a-38108b9250f5.png)|![Screenshot_20220120_135623](https://user-images.githubusercontent.com/1848238/150353639-86dc4e4c-86ec-4857-804e-2f2ff9261e2e.png)

**Comment by waclaw66** [comment]
Thanks for the investigation. I think it's totaly torelable now, the only disturbance is the counter. It's much better that before 1.3.7. The issue can be closed.

**Comment by ouchadam** [comment]
thanks for all the help! I'll close this ticket off as a _won't fix_ but we can definitely reopen if we find more OEMs/roms with the same issue

If anyone is finding this really troublesome a workaround is to disable the "conversation grouping" notification settings, although it means losing the benefit of conversations 

| TOGGLE | RESULT | 
| --- | --- |
![2022-01-21T09:21:25,388677615+00:00](https://user-images.githubusercontent.com/1848238/150501220-081c208e-50e5-4e46-a693-aec50cde1aac.png)|![2022-01-21T09:23:54,169369329+00:00](https://user-images.githubusercontent.com/1848238/150501506-4e659f6a-b786-49a1-86e1-62ac5512609e.png)
